Fix #20088: Behaviour of voiceover editor of explorations

sissamrcorreia opened the pull request:
## Overview

1. This PR fixes #20088.
2. This PR does the following: Address both reported issues. (a) Regarding English lessons, the solution resolves cases where translation placeholders incorrectly appeared despite the content already being in English. The updated system compares the active translation language with the exploration's displayed language, ensuring translation-related elements only appear when there is an actual language difference. (b) For non-English languages, the changes eliminate the incorrect display of English text as the base for voiceovers. The getRequiredHtml method now returns null when no translation exists for the target language, preventing misleading references to unavailable content.
Additional improvements feature enhanced visual feedback through color-coded bars: gray indicates no action is needed (either because no text is available for voiceover or because the translation has not been created yet), while red highlights sections requiring voiceovers (where text exists but lacks corresponding voiceover). The implementation includes clearer user messaging, specifically displaying "There is no text available to voiceover" when appropriate. Progress tracking now accurately reflects voiceover statuses, significantly simplifying identification and prioritization of pending work throughout the localization workflow.
3. The original bug occurred because: (a) The bug occurred because the system failed to recognize content already in English, leading to unnecessary translation placeholders. (b) The bug happened when English text was used as the base for voiceovers in non-English languages, even without available translations, causing misleading references. In both cases, the progress bar and tab colors incorrectly indicated sections as needing action, adding to user confusion.

## Essential Checklist

Please follow the [instructions for making a code change](https://github.com/oppia/oppia/wiki/Make-a-pull-request).

- [x] I have linked the issue that this PR fixes in the "Development" section of the sidebar.
- [x] I have checked the "Files Changed" tab and confirmed that the changes are what I want to make.
- [x] I have written tests for my code.
- [x] The **PR title** starts with "Fix #bugnum: " or "Fix part of #bugnum: ...", followed by a short, clear summary of the changes.
- [x] I have assigned the correct reviewers to this PR (or will leave a comment with the phrase "@{{reviewer_username}} PTAL" if I can't assign them directly).


## Proof that changes are correct

https://github.com/user-attachments/assets/c826e48c-5f7b-4451-b98a-7886b417ee64

https://github.com/user-attachments/assets/09b2da3e-629b-4b65-adc3-a7348ce3e542

## PR Pointers

- Never force push! If you do, your PR will be closed.
- To reply to reviewers, follow these instructions: https://github.com/oppia/oppia/wiki/Make-a-pull-request#step-5-address-review-comments-until-all-reviewers-approve
- Some e2e tests are flaky, and can fail for reasons unrelated to your PR. We are working on fixing this, but in the meantime, if you need to restart the tests, please check the ["If your build fails" wiki page](https://github.com/oppia/oppia/wiki/If-CI-checks-fail-on-your-PR).
- See the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia's-code-owners-and-checks-to-be-carried-out-by-developers) for what code owners will expect.

sissamrcorreia committed:
Fix: Behaviour of voiceover editor of explorations

- Fixes translation placeholder display in English lessons by
  comparing active and exploration languages.
- Corrects base text display for non-English voiceovers, returning
  null if no translation exists.
- Adds color-coded feedback bars (gray/red) for voiceover status
  and clearer "no text available" messages.
- Fixes progress tracking for pending voiceovers.
- Updated the existing tests to match the new expected behaviours.

Nik-09 reviewed (changes requested):
Hi @sissamrcorreia, could you please fix the CI failing CI checks?

seanlip committed:
Merge branch 'develop' into fix-20088

sissamrcorreia commented:
I am currently working on fixing the failing CI checks. I have encountered some issues while running the tests locally, but I am troubleshooting them now. I will try to update the PR soon.

seanlip commented on core/templates/pages/exploration-editor-page/translation-tab/services/translation-status.service.ts:
```diff
@@ -46,6 +47,7 @@ export class TranslationStatusService implements OnInit {
   ALL_ASSETS_AVAILABLE_COLOR: string = '#16A765';
   FEW_ASSETS_AVAILABLE_COLOR: string = '#E9B330';
   NO_ASSETS_AVAILABLE_COLOR: string = '#D14836';
+  NULL_ASSETS_AVAILABLE_COLOR: string = '#6E6E6E';
```
SOURCE_CONTENT_IS_EMPTY_COLOR (perhaps) -- otherwise this is too similar to NO_ASSETS_AVAILABLE_COLOR.

Feel free to suggest a different naming convention for these colours if there is an alternative you can think of that would convey the meaning more clearly.

sissamrcorreia replied:
Done!

seanlip commented on core/templates/pages/exploration-editor-page/translation-tab/services/translation-status.service.ts:
```diff
@@ -187,10 +190,34 @@ export class TranslationStatusService implements OnInit {
+            stateName
+          );
+
+        if (
```
What happens when this condition doesn't hold? I am a bit confused about why it depends on whether the voiceover mode is active; shouldn't this just rely on what's in the exploration and what voiceovers are present?

seanlip commented on core/templates/pages/exploration-editor-page/translation-tab/state-translation/state-translation.component.ts:
```diff
@@ -137,26 +141,38 @@ export class StateTranslationComponent implements OnInit, OnDestroy {
 
     let langCode = this.translationLanguageService.getActiveLanguageCode();
+
+    if (langCode == this.explorationLanguageCodeService.displayed) {
```
You define needsTranslation() above for this; why not use it?

seanlip commented on core/templates/pages/exploration-editor-page/translation-tab/state-translation/state-translation.component.ts:
```diff
@@ -137,26 +141,38 @@ export class StateTranslationComponent implements OnInit, OnDestroy {
       )
     ) {
-      return subtitledHtml.html;
+      return null;
```
Why did this (and the following) change?

sissamrcorreia replied:
Previously (line 151), the system returned untranslated HTML content when no translation was available. This meant displaying the exploration's original language content even when viewed in another language, which could confuse users.
The new behaviour (line 167) now returns null instead, only showing:
"The translation for this section has not been created yet. Switch to translation mode to add a text translation."

Do you have suggestions for alternative ways to handle this case?

seanlip replied:
@sissamrcorreia Thanks for the explanation. This might be OK but it needs comments in the code, a docstring written for the function, and possibly a change to the name to more clearly indicate what the function is returning.

(I.e. I don't think the change is necessarily wrong -- it just needs to be communicated properly so that future developers reading this aren't confused. Thanks!)

seanlip commented on core/templates/pages/exploration-editor-page/translation-tab/state-translation/state-translation.component.ts:
```diff
@@ -523,6 +542,53 @@ export class StateTranslationComponent implements OnInit, OnDestroy {
+      this.needsTranslation() &&
+      this.translationTabActiveModeService.isVoiceoverModeActive()
+    ) {
+      return {
```
Reduce the duplication and branching. Compute the colour first, then have a single return statement that makes use of it.

seanlip commented on core/templates/pages/exploration-editor-page/translation-tab/state-translation/state-translation.component.ts:
```diff
@@ -523,6 +542,53 @@ export class StateTranslationComponent implements OnInit, OnDestroy {
+    };
+  }
+
+  tabStatusColorStyleContent(tabId: string): object {
```
It would be better to name this function so that it starts with a verb. Feel free to do the same for other functions as well.

sissamrcorreia replied:
Done!

sissamrcorreia committed:
Refactor: Voiceover tab color styles and retrieval logic

- Update methods to use unified getTabStatusColorStyle, and rename constants for clarity.
- Updated the existing tests to match the new expected behaviours.

seanlip reviewed (changes requested):
@sissamrcorreia Just a note, please see the [rules for making PRs](https://github.com/oppia/oppia/wiki/Rules-for-making-PRs#step-5-address-review-comments-until-all-reviewers-approve), particularly step 5. In particular:

- Please ensure that you leave a response to all comments. Don't resolve them, since that makes it hard to tell whether a comment has been addressed properly or not.
- If you need a PR/question looked at, please update the list of assignees to ensure that it clearly reflects who needs to take the next action. Otherwise responses will likely be delayed since reviewers check github.com/pulls/assigned to see what they need to focus on, and if they aren't in your PRs assignee list then they won't see your PR.

Thanks!

sissamrcorreia committed:
Fix: Formatting error

sissamrcorreia commented:
Hi @seanlip!

I have been working on this, but it turned out to be more complex than I expected, especially since this is my very first issue, and I am still getting familiar with the process.

I do not want to block progress, so it might be best for someone else to take over this issue. In the meantime, I will look for other ways to contribute. Really appreciate your understanding and guidance.

Nik-09 commented:
Hi @sissamrcorreia, if you can point to the specific errors/issues/challenges that you are facing, then we can help you to get that sorted.
Thanks!

sissamrcorreia commented:
Hi @Nik-09,
I really appreciate your willingness to help! After spending more time on this, I have realized that the main challenge lies in how closely the voiceover and translation logic are intertwined. 
While trying to address the original issue, I noticed that changes meant to fix voiceovers inadvertently affected translation functionality. Five previously passing tests in ```translation-status.service.spec.ts``` are now failing.

The interactions between these systems are tricky to trace, and I am still getting comfortable with how they work together. 

Do you have suggestions on how to better understand all possible workflows and edge cases? 

Thank you in advance.
